inc/zoo/map/RobinHood_util.h

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include "zoo/swar/SWAR.h"
+
+#include <array>
+#include <cassert>
+#include <vector>
+
+namespace zoo {
+
+namespace rh {
+
+using u64 = uint64_t;
+using u32 = uint32_t;
+using u16 = uint16_t;
+using u8 = uint8_t;
+
+template<int NBits>
+constexpr auto cheapOkHash(u64 n) noexcept {
+    constexpr auto shift = (NBits * ((64 / NBits)-1));
+    constexpr u64 allOnes = meta::BitmaskMaker<u64, 1, NBits>::value;
+    auto temporary = allOnes * n;
+    auto higestNBits = temporary >> shift;
+    return (0 == (64 % NBits)) ? 
+        higestNBits : swar::isolate<NBits, u64>(higestNBits);
+}
+
+/// Does some multiplies with a lot of hash bits to mix bits, returns only a few
+/// of them.
+template<int NBits> auto badMixer(u64 h) noexcept {
+    constexpr u64 allOnes = ~0ull;
+    constexpr u64 mostSigNBits = swar::mostNBitsMask<NBits, u64>();
+    auto tmp = h * allOnes;
+    
+    auto mostSigBits = tmp & mostSigNBits;
+    return mostSigBits >> (64 - NBits);
+}
+
+template<typename Key, typename T = u64> T fibhash(Key k) noexcept {
+    constexpr auto FibonacciGoldenRatioReciprocal64 = 11400714819323198485ull;
+    constexpr auto FibonacciGoldenRatioReciprocal32 = 2654435769u;
+    constexpr auto FibonacciGoldenRatioReciprocal16 = 40503u;
+    constexpr auto FibonacciGoldenRatioReciprocal8 = 159u;
+    if constexpr(sizeof(Key) == 8) {
+        u64 r = k * FibonacciGoldenRatioReciprocal64;
+        return r; 
+    } else if constexpr(sizeof(Key) == 4) {
+        u32 r = k * FibonacciGoldenRatioReciprocal32; 
+        return r;
+    } else if constexpr(sizeof(Key) == 2) {
+        u16 r = k * FibonacciGoldenRatioReciprocal16;
+        return r;
+    } else if constexpr(sizeof(Key) == 1) {
+        u8 r = k * FibonacciGoldenRatioReciprocal8;
+        return r;
+    }
+    assert(false);
+}
+
+/// Evenly map a large int to an int without division or modulo.
+template<int SizeTable, typename T>
+constexpr int mapToSlotLemireReduction(T halved) {
+    return (halved * T(SizeTable)) >> (sizeof(T)/2); 
+}
+
+
+template<typename Key>auto reducedhashUnitary(Key k) noexcept {
+    return 1;
+}
+
+template<typename Key>auto slotFromKeyUnitary(Key k) noexcept {
+    return 1;
+}
+
+template<int NBitsHash, int NBitsPSL, typename T = u64> struct SlotOperations {
+    using SSL = swar::SWARWithSubLanes<NBitsHash, NBitsPSL, T>;
+    using SM = swar::SWAR<NBitsHash+ NBitsPSL, T>;
+    using BoolSM = swar::BooleanSWAR<NBitsHash+ NBitsPSL, T>;
```

Comment:
I object to this name.
It seems this class is used as a namespace (not the syntactical feature but the concept).